kotlin/src/main/com/looker/rtl/SdkResult.kt

```diff
@@ -0,0 +1,104 @@
+package com.looker.rtl
+
+import io.ktor.client.call.receive
+import io.ktor.client.response.HttpResponse
+import io.ktor.http.isSuccess
+import kotlinx.coroutines.runBlocking
+
+class FailureResponseError(val result: SdkResult<*, *>) : Exception()
+
+interface SdkResponse<T> {
+    val success: Boolean
+    val statusCode: Int
+    val method: HttpMethod
+    val path: String
+    fun body(): T
+}
+
+sealed class SdkResult<out TSuccess, out TFailure> {
```

Comment:
I know you're just augmenting the existing pattern of an sdk method returns a "meta" object that you can get the actual SDK response type from (on success) with `.body()` or use `.ok()` to return that or throw. And this structure seems reasonable to me.

My question is at a higher design level for you and @jkaster - searching through some pretty popular python sdks (github, azure, google cloud sdks) - the idiomatic approach for these python sdks is for a method to behave as if the `sdk.ok()` (or `response.ok()` here) method is automatically called (i.e. the sdk method either raises or returns the `TSuccess` object directly)

For python I'd actually propose adding a method on the SDK response types that exposes the success/status_code/method/path/raw_body metadata (and for the error case throwing an exception that has this same metadata access method).

Is it ok from our design perspective that the python sdk behaves so differently? We could make it behave like the others but then I think we lose the idiomatic nature that other popular python sdk libraries are following.

thoughts?

Reply:
It's definitely simpler if the sdk response is always just TSuccess or an exception.  I guess since my motivation for this was better error handling, we could limit the extra functionality to just the error path.  That would limit the client's ability to introspect on the response in the success case, but maybe we don't care about that.  Also, I remember @jkaster talking about how he didn't want to throw in the failure/error case until you actually went to get the value.  Something about working better in an async flow, so maybe we should look at what node or something else natively async does here.

I'd be up for a quick group chat about this topic.

Reply:
happy to participate - I'll let you and our fearless leader @jkaster coordinate :-)